File: backend/app/schemas/analytics.py
"""求职数据看板（R-14）的响应结构。

只承载结构，**所有计数口径的唯一来源是 ``models/tracker.py`` 的 ``STATUSES`` /
``STATUS_RANK`` / ``STATUS_LABELS``**，看板不另写枚举。具体口径定义与计算见
``services/analytics.py`` 的模块 docstring（那里也写明了**因数据模型不支持而刻意不提供**
的指标，别再提）。

字段全部带默认值：空库时每项都是明确的零/空列表，**不是 None**——前端据此渲染空态，
不需要到处判空。新增字段一律与 ``frontend/src/types/analytics.ts`` 逐字对应（snake_case）。
"""
from pydantic import BaseModel, Field


class FunnelStage(BaseModel):
    """漏斗里的一个阶段：状态键、面向用户的中文名与计数。"""

    status: str
    label: str
    count: int


class TrendPoint(BaseModel):
    """趋势图上的一个点：一个自然月的投递数量。"""

    month: str = ""  # YYYY-MM
    label: str = ""  # 面向展示的短标签（如 "9月"）
    count: int = 0


class LabeledCount(BaseModel):
    """带展示名的分桶计数（周内分布 / 公司排行 / 记录来源共用）。

    ``label`` 只由**拥有该词表**的一侧填：``SOURCE_LABELS`` 与公司名在后端，
    所以这里给得出；内推状态的中文名归前端（见 :class:`KeyedCount`）。
    """

    key: str = ""
    label: str = ""
    count: int = 0


class KeyedCount(BaseModel):
    """只有键与计数的分桶（内推状态分布用）。

    **刻意不带 ``label``**：内推状态的展示名归前端 ``REFERRAL_STATUS_LABELS``，
    后端再写一份就是同一件事的两份定义、迟早漂移。
    """

    key: str = ""
    count: int = 0


class AppliedGapOut(BaseModel):
    """投递日期的填写情况，供前端如实说明"这张图为什么是空的"。

    趋势图与周内分布只统计有 ``applied_at`` 的记录；没有这组数字，一张全零的图会被
    读成"这几个月真的一份没投"，而事实是那些记录**没填日期**。
    """

    dated: int = 0
    undated: int = 0
    total: int = 0


class ReferralBlockOut(BaseModel):
    """内推转化（口径来自 ``referral_service``，看板只搬运）。"""

    total: int = 0
    converted: int = 0
    rate: float = 0.0


class ReminderCountsOut(BaseModel):
    """待办提醒按紧急度分档（阈值口径来自 ``reminder_service.reminder_urgency``）。"""

    overdue: int = 0
    soon: int = 0
    upcoming: int = 0
    later: int = 0
    total: int = 0


class DashboardOut(BaseModel):
    """看板整体响应：概览指标 + 四个主题区块所需的分组数据。"""

    # ===== 概览 =====
    total_applications: int = 0
    valid_applications: int = 0
    interview_count: int = 0
    interview_rate: float = 0.0
    assessment_count: int = 0
    assessment_to_interview_count: int = 0
    assessment_pass_rate: float = 0.0
    offer_count: int = 0
    offer_rate: float = 0.0
    funnel: list[FunnelStage] = Field(default_factory=list)
    trend: list[TrendPoint] = Field(default_factory=list)

    # ===== 转化与卡点 =====
    active_count: int = 0
    stalled_count: int = 0
    no_next_action_count: int = 0

    # ===== 时间与节奏 =====
    weekday: list[LabeledCount] = Field(default_factory=list)
    applied_date_gap: AppliedGapOut = Field(default_factory=AppliedGapOut)
    recent_7d_count: int = 0
    recent_30d_count: int = 0
    reminder_counts: ReminderCountsOut = Field(default_factory=ReminderCountsOut)

    # ===== 渠道与去向 =====
    top_companies: list[LabeledCount] = Field(default_factory=list)
    other_company_count: int = 0
    record_sources: list[LabeledCount] = Field(default_factory=list)
    referral: ReferralBlockOut = Field(default_factory=ReferralBlockOut)
    referral_status: list[KeyedCount] = Field(default_factory=list)

    # ===== 简历与健康度 =====
    resume_count: int = 0
    resume_scanned_count: int = 0
    resume_with_warnings_count: int = 0
    resume_with_placeholders_count: int = 0
    unverified_claim_count: int = 0
    track_resume_linked_count: int = 0


__all__ = [
    "AppliedGapOut",
    "DashboardOut",
    "FunnelStage",
    "KeyedCount",
    "LabeledCount",
    "ReferralBlockOut",
    "ReminderCountsOut",
    "TrendPoint",
]

File: frontend/src/types/analytics.ts
/** 求职数据看板（后端 /api/analytics）。
 *
 * 字段与 `backend/app/schemas/analytics.py` 的 `DashboardOut` **逐字对应**（snake_case，
 * 没有 camelCase 转换层）。各指标的口径写在 `services/analytics.py` 的模块 docstring 里，
 * 那里也写明了因数据模型不支持而**刻意不提供**的指标。
 *
 * 所有字段在后端都有默认值：空库时是明确的零/空数组，**不是 null**——所以这里不需要
 * 到处判空，直接渲染空态即可。
 */

export interface FunnelStage {
  status: string;
  label: string;
  count: number;
}

export interface TrendPoint {
  /** YYYY-MM。 */
  month: string;
  /** 面向展示的短标签，如 "9月"。 */
  label: string;
  count: number;
}

/** 带展示名的分桶计数（周内分布 / 公司排行 / 记录来源共用）。 */
export interface LabeledCount {
  key: string;
  label: string;
  count: number;
}

/** 只有键与计数的分桶（内推状态分布）。中文名由前端 `REFERRAL_STATUS_LABELS` 提供。 */
export interface KeyedCount {
  key: string;
  count: number;
}

/** 投递日期的填写情况：趋势图与周内分布**只统计有日期的记录**，没有这组数字，
 *  一张全零的图会被读成"这几个月真的一份没投"。 */
export interface AppliedGap {
  dated: number;
  undated: number;
  total: number;
}

export interface ReferralBlock {
  total: number;
  converted: number;
  rate: number;
}

export interface ReminderCounts {
  overdue: number;
  soon: number;
  upcoming: number;
  later: number;
  total: number;
}

export interface AnalyticsDashboard {
  // ===== 概览 =====
  total_applications: number;
  valid_applications: number;
  interview_count: number;
  interview_rate: number;
  assessment_count: number;
  assessment_to_interview_count: number;
  assessment_pass_rate: number;
  offer_count: number;
  offer_rate: number;
  funnel: FunnelStage[];
  trend: TrendPoint[];

  // ===== 转化与卡点 =====
  active_count: number;
  /** 进行中且**超过 7 天没有任何更新**（改一条备注也算更新），不叫"面试卡住"。 */
  stalled_count: number;
  no_next_action_count: number;

  // ===== 时间与节奏 =====
  weekday: LabeledCount[];
  applied_date_gap: AppliedGap;
  /** 滚动窗口（按 created_at），与用户时区无关。 */
  recent_7d_count: number;
  recent_30d_count: number;
  reminder_counts: ReminderCounts;

  // ===== 渠道与去向 =====
  top_companies: LabeledCount[];
  other_company_count: number;
  /** **记录怎么进系统的**，不是投递渠道。 */
  record_sources: LabeledCount[];
  referral: ReferralBlock;
  referral_status: KeyedCount[];

  // ===== 简历与健康度 =====
  resume_count: number;
  /** 健康度扫描窗口内的份数；小于 `resume_count` 时说明只扫了最近这批。 */
  resume_scanned_count: number;
  resume_with_warnings_count: number;
  resume_with_placeholders_count: number;
  unverified_claim_count: number;
  /** 投递-简历关联覆盖率。录入界面目前没有关联入口，所以它是用来**暴露缺口**的。 */
  track_resume_linked_count: number;
}
